### ISSUE TITLE
Stream.filterMapEffect but accepting Effect<Option<A>> instead of Option<Effect<A>>

### ISSUE THREAD

**Issue by Schniz** [Effect-TS/effect#3018]
### What is the problem this feature would solve?

`Stream.filterMapEffect` is unique to the rest of the `Stream.*Effect` functions: it expects an effect wrapped in another data structure instead of an effect _containing_ a data structure: `(a: A) => Option<Effect<B>>`

or in other words, if you want to `filterMapEffect` you can't filter map based on an effect value, but instead you can choose whether to return an effect or not

Let's say we want to map a stream of numbers into their square roots:

```typescript
declare const sqrt = (number: number) => Effect.Effect<Option<number>>;

const numbers = Stream.fromIterable([-1,2,3])

// What we can do right now
const squareRoots_ = numbers.pipe(
  Stream.mapEffect(number => sqrt(number)),
  Stream.filterMap(opt => opt)
);

// With this filterMapEffectX (name tbd):
const squareRoots = numbers.pipe(
  Stream.filterMapEffectX(number => sqrt(number))
)
```

### What is the feature you are proposing to solve the problem?

it seems counter intuitive that this is the signature IMO, and I know that we probably can't break this API as Effect 3 is stable -- so we'll have to think of a different way to call it.

This is obviously can be done in userland as I mentioned previously, which is how I implemented it on my projects:

```typescript
import { Function, Stream, Effect, Option } from 'effect';

export const filterMapEffect = Function.dual<
  <A, A2, E2, R2>(
    fn: (a: A) => Effect.Effect<Option.Option<A2>, E2, R2>,
  ) => <E, R>(
    stream: Stream.Stream<A, E, R>,
  ) => Stream.Stream<A2, E2 | E, R2 | R>,
  <A, E, R, A2, E2, R2>(
    stream: Stream.Stream<A, E, R>,
    fn: (a: A) => Effect.Effect<Option.Option<A2>, E2, R2>,
  ) => Stream.Stream<A2, E2 | E, R2 | R>
>(2, (stream, fn) =>
  Stream.filterMap(Stream.mapEffect(stream, fn), Function.identity),
);
```

### What alternatives have you considered?

The alternative is to simply keep that a userland implementation. But I think this API makes sense in the "standard library".

**Comment by datner** [comment]
I am not sure if it is unique, I'll leave that judgement to others. I do think that there is logic in the methods. Though it's description is incorrect. It does not effectfully filter and maps, it filters and effectfully maps.


Regarding usefulness......... I tend to agree, though this util is quite old and at the time of its making there was literally no util in effect that could effectfully filter. Filtering was an explicitly sync operation. Not justifying it, but maybe thats the background for it 🤷🏻

**Comment by Schniz** [comment]
hey hey i'm not pointing fingers <picture data-single-emoji=":relieved_galactic:" title=":relieved_galactic:"><img class="emoji" width="20" height="auto" src="https://emoji.slack-edge.com/T0CAQ00TU/galactic_relifed/33851380ad8f06bb.png" alt=":relieved_galactic:" align="absmiddle"></picture> just want to make it better!

**Comment by datner** [comment]
@mikearnaldi I think you might have accidentally closed this issue despite reverting #3023
